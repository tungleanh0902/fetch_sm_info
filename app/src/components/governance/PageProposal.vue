<template>
  <TmPage data-title="Proposal" :loading="$apollo.queries.proposal.loading">
    <TmDataNotFound v-if="!found" />
    <div v-else class="proposal">
      <ProposalHeader
        :proposal="proposal"
        :status="status"
        @open-vote-modal="onVote"
        @open-deposit-modal="onDeposit"
      />

      <ProposalStatusBar
        v-if="tallyHasValues"
        :status="status"
        :status-begin-time="proposal.statusBeginTime"
        :total-votes="proposal.tally.total"
        :proposal="proposal"
      />

      <ParticipantList
        v-if="participants"
        :participants="participants"
        :show-amounts="true"
      />

      <template v-if="proposal.detailedVotes.timeline.length">
        <Timeline :timeline="proposal.detailedVotes.timeline" />
      </template>

      <ProposalDescription
        :proposal="proposal"
        :supporting-links="proposal.detailedVotes.links"
      />

      <ModalDeposit
        v-if="status.value === governanceStatusEnum.DEPOSITING"
        ref="modalDeposit"
        :proposal-id="
          currentNetwork.network_type === `polkadot`
            ? proposal.proposalId
            : proposalId
        "
        :proposal-title="proposal.title || ''"
        :denom="parameters.depositDenom || currentNetwork.stakingDenom"
        :deposits="proposal.detailedVotes.deposits"
        @success="() => afterDeposit()"
      />
      <ModalVotePolkadot
        v-if="
          currentNetwork.network_type === `polkadot` &&
          status.value === governanceStatusEnum.VOTING
        "
        ref="modalVote"
        :proposal-id="proposal.proposalId"
        :proposal-title="proposal.title || ''"
        :last-vote-option="vote"
        @success="() => afterVote()"
      />
      <ModalVote
        v-else
        ref="modalVote"
        :proposal-id="proposalId"
        :proposal-title="proposal.title || ''"
        :last-vote-option="vote"
        :number-of-seconds="
          isPolkadotDemocracy ? Number(proposal.detailedVotes.votesSum) : 0
        "
        @success="() => afterVote()"
      />
    </div>
  </TmPage>
</template>

<script>
import { mapGetters } from "vuex"
import { percent, prettyInt } from "scripts/num"
import { governanceStatusEnum } from "scripts/proposal-status"
import { date, fromNow } from "src/filters"
import TmDataNotFound from "common/TmDataNotFound"
import ModalDeposit from "src/ActionModal/components/ModalDeposit"
import ModalVote from "src/ActionModal/components/ModalVote"
import ModalVotePolkadot from "src/ActionModal/components/ModalVotePolkadot"
import TmPage from "common/TmPage"
import ParticipantList from "governance/ParticipantList"
import ProposalHeader from "governance/ProposalHeader"
import ProposalStatusBar from "governance/ProposalStatusBar"
import ProposalDescription from "governance/ProposalDescription"
import Timeline from "governance/Timeline"
import { getProposalStatus } from "scripts/proposal-status"
import { ProposalItem, GovernanceParameters, Vote } from "src/gql"
import BigNumber from "bignumber.js"
import gql from "graphql-tag"

export default {
  name: `page-proposal`,
  components: {
    ModalDeposit,
    ModalVote,
    ModalVotePolkadot,
    TmDataNotFound,
    TmPage,
    ParticipantList,
    ProposalHeader,
    ProposalDescription,
    ProposalStatusBar,
    Timeline,
  },
  filters: {
    prettyInt,
    percent,
    date,
    fromNow,
    lowerCase: (text) => (text ? text.toLowerCase() : ""),
  },
  data: () => ({
    proposals: [],
    vote: undefined,
    proposal: {
      status: "",
      proposer: "",
      tally: {},
      validator: {},
    },
    parameters: {
      depositDenom: "",
    },
    error: undefined,
    found: false,
    loaded: false,
    governanceStatusEnum,
  }),
  computed: {
    ...mapGetters([`address`, `network`, `currentNetwork`]),
    proposalId() {
      return this.$route.params.proposalId
    },
    status() {
      return getProposalStatus(this.proposal)
    },
    noVotes() {
      return BigNumber(this.proposal.tally.total).eq(0)
    },
    tallyHasValues() {
      return Object.values(this.proposal.tally)
        .filter((value) => value !== `Tally`)
        .find((value) => value)
    },
    isPolkadotDemocracy() {
      return (
        this.currentNetwork.network_type === `polkadot` &&
        this.status.value === governanceStatusEnum.DEPOSITING
      )
    },
    participants() {
      if (
        this.proposal.detailedVotes.votes &&
        this.proposal.detailedVotes.votes.length > 0
      ) {
        return this.proposal.detailedVotes.votes.map((vote) => ({
          ...vote.voter,
          amount: vote.amount,
          option: vote.option,
        }))
      } else if (
        this.proposal.detailedVotes.deposits &&
        this.proposal.detailedVotes.deposits.length > 0
      ) {
        // a bit hacky but working
        return this.proposal.detailedVotes.deposits.map((deposit) => ({
          ...deposit.depositer,
          amount: deposit.amount[0],
        }))
      }
      return undefined
    },
  },
  methods: {
    onVote() {
      this.$refs.modalVote.open()
    },
    afterVote() {
      this.$apollo.queries.vote.refetch()
    },
    onDeposit() {
      this.$refs.modalDeposit.open()
    },
    afterDeposit() {
      this.$apollo.queries.proposal.refetch()
    },
  },
  apollo: {
    proposal: {
      /* istanbul ignore next */
      query() {
        return ProposalItem(this.network)
      },
      /* istanbul ignore next */
      update(data) {
        this.loaded = true
        if (data.proposal) this.found = true
        return data.proposal || {}
      },
      /* istanbul ignore next */
      variables() {
        return {
          id: this.proposalId,
        }
      },
      /* istanbul ignore next */
      result(data) {
        if (data.proposal) this.found = true
        this.error = data.error
      },
    },
    parameters: {
      /* istanbul ignore next */
      query() {
        return GovernanceParameters(this.network)
      },
      /* istanbul ignore next */
      update(data) {
        return data.governanceParameters || {}
      },
      /* istanbul ignore next */
      skip() {
        // only Tendermint networks have this network-wide "governance parameters" logic
        return this.currentNetwork.network_type !== `cosmos`
      },
      /* istanbul ignore next */
      result(data) {
        this.error = data.error
      },
    },
    vote: {
      /* istanbul ignore next */
      query() {
        return Vote(this.network)
      },
      /* istanbul ignore next */
      variables() {
        return {
          proposalId: this.proposalId,
          address: this.address,
        }
      },
      /* istanbul ignore next */
      skip() {
        return !this.address || !this.found
      },
      update(data) {
        if (data.vote) return data.vote.option
        return undefined
      },
      /* istanbul ignore next */
      result(data) {
        this.error = data.error
      },
    },
    $subscribe: {
      blockAdded: {
        /* istanbul ignore next */
        variables() {
          return {
            networkId: this.network,
          }
        },
        /* istanbul ignore next */
        query() {
          return gql`
            subscription($networkId: String!) {
              blockAdded(networkId: $networkId) {
                height
              }
            }
          `
        },
        /* istanbul ignore next */
        skip() {
          return !this.found
        },
        /* istanbul ignore next */
        result() {
          if (
            // Don't update passed or rejected proposals
            this.proposal.status !== "Passed" &&
            this.proposal.status !== "Rejected" &&
            this.loaded
          ) {
            this.$apollo.queries.proposal.refetch()
            this.$apollo.queries.parameters.refetch()
            this.$apollo.queries.vote.refetch()
          }
        },
      },
    },
  },
}
</script>
<style scoped>
.proposal {
  padding: 0 1rem;
}
</style>
