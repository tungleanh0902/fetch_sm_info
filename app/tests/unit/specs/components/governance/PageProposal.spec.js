"use strict"

import PageProposal from "governance/PageProposal"
import { createLocalVue, shallowMount } from "@vue/test-utils"
import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { proposals } from "../../store/json/proposals"

describe(`PageProposal`, () => {
  let wrapper, $store

  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.use(Vuelidate)
  localVue.directive(`tooltip`, () => {})
  localVue.directive(`focus`, () => {})

  const state = {
    session: {
      address: `cosmos1xxxx`,
    },
  }

  const getters = {
    connected: true,
    address: `cosmos1xxxx`,
    currentNetwork: {
      id: "cosmos-hub-mainnet",
      stakingDenom: `ATOM`,
      network_type: `cosmos`,
    },
  }
  let args

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      getters,
      state,
    }

    const $apollo = {
      queries: {
        proposals: {
          loading: false,
          error: undefined,
        },
        proposal: {
          loading: false,
          error: undefined,
        },
        parameters: {
          loading: false,
          error: undefined,
        },
        vote: {
          loading: false,
          error: undefined,
        },
      },
    }
    args = {
      localVue,
      propsData: {
        proposalId: `33`,
      },
      mocks: {
        $store,
        $apollo,
        $route: {
          params: { proposalId: 19 },
        },
      },
      stubs: [`router-link`],
    }
    wrapper = shallowMount(PageProposal, args)
    wrapper.setData({
      proposal: proposals[2],
      found: true,
    })
  })

  it("should show a loader if the necessary data hasen't been loaded", () => {
    wrapper.vm.$apollo.queries.proposal.loading = true
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`shows an error if the proposal couldn't be found`, () => {
    wrapper = shallowMount(PageProposal, {
      ...args,
      propsData: { proposalId: `666` },
    })
    wrapper.setData({ error: { message: "Error" } })
    expect(wrapper.element).toMatchSnapshot()
  })

  describe(`Proposal status`, () => {
    it(`displays correctly a proposal that 'Passed'`, () => {
      wrapper.vm.proposal.status = `Passed`
      expect(wrapper.vm.status).toMatchObject({
        title: `Passed`,
        value: `PASSED`,
        color: `green`,
      })
    })

    it(`displays correctly a 'Rejected' proposal`, () => {
      wrapper.vm.proposal.status = `Rejected`
      expect(wrapper.vm.status).toMatchObject({
        title: `Rejected`,
        value: `REJECTED`,
        color: `red`,
      })
    })

    it(`displays correctly a proposal on 'DepositPeriod'`, () => {
      wrapper.vm.proposal.status = `DepositPeriod`
      expect(wrapper.vm.status).toMatchObject({
        title: `Deposit Period`,
        value: `DEPOSITING`,
        color: `orange`,
      })
    })

    it(`displays correctly a proposal on 'VotingPeriod'`, () => {
      wrapper.vm.proposal.status = `VotingPeriod`
      expect(wrapper.vm.status).toMatchObject({
        title: `Voting Period`,
        color: `highlight`,
      })
    })

    it(`shows error status`, () => {
      wrapper.vm.proposal.status = ``
      expect(wrapper.vm.status).toMatchObject({
        title: `Error`,
        color: `grey`,
      })
    })
  })

  describe(`Modal onVote`, () => {
    it(`shows the last valid vote`, async () => {
      wrapper.setData({ vote: "Yes" })
      expect(wrapper.html()).toMatchSnapshot()
    })

    it(`refetches user vote after a successful voting`, () => {
      wrapper.vm.$apollo.queries.vote.refetch = jest.fn()
      wrapper.vm.afterVote()
      expect(wrapper.vm.$apollo.queries.vote.refetch).toHaveBeenCalled()
    })
  })

  describe(`Modal onDeposit`, () => {
    it(`opens deposit modal`, async () => {
      const self = {
        $refs: { modalDeposit: { open: jest.fn() } },
        proposalId: `666`,
      }

      await PageProposal.methods.onDeposit.call(self)

      expect(self.$refs.modalDeposit.open).toHaveBeenCalled()
    })

    it(`refetches proposal data after a successful depositing`, () => {
      wrapper.vm.$apollo.queries.proposal.refetch = jest.fn()
      wrapper.vm.afterDeposit()
      expect(wrapper.vm.$apollo.queries.proposal.refetch).toHaveBeenCalled()
    })
  })
})
