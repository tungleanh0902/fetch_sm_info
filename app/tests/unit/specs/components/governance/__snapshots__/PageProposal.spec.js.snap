// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageProposal Modal onVote shows the last valid vote 1`] = `
<tmpage-stub loaderpath="" emptytitle="No data" emptysubtitle="" data-title="Proposal" proposalid="33">
  <div class="proposal">
    <proposalheader-stub proposal="[object Object]" status="[object Object]"></proposalheader-stub>
    <proposalstatusbar-stub status="[object Object]" statusbegintime="2019-08-25T08:46:16.394157593Z" proposal="[object Object]" total-votes="0"></proposalstatusbar-stub>
    <!---->
    <timeline-stub timeline="[object Object]"></timeline-stub>
    <proposaldescription-stub proposal="[object Object]" supportinglinks=""></proposaldescription-stub>
    <!---->
    <modalvote-stub proposalid="19" proposaltitle="test" number-of-seconds="0" lastvoteoption="Yes"></modalvote-stub>
  </div>
</tmpage-stub>
`;

exports[`PageProposal should show a loader if the necessary data hasen't been loaded 1`] = `
<tmpage-stub
  data-title="Proposal"
  emptysubtitle=""
  emptytitle="No data"
  loaderpath=""
  loading="true"
  proposalid="33"
>
  <div
    class="proposal"
  >
    <proposalheader-stub
      proposal="[object Object]"
      status="[object Object]"
    />
     
    <proposalstatusbar-stub
      proposal="[object Object]"
      status="[object Object]"
      statusbegintime="2019-08-25T08:46:16.394157593Z"
      total-votes="0"
    />
     
    <!---->
     
    <timeline-stub
      timeline="[object Object]"
    />
     
    <proposaldescription-stub
      proposal="[object Object]"
      supportinglinks=""
    />
     
    <!---->
     
    <modalvote-stub
      number-of-seconds="0"
      proposalid="19"
      proposaltitle="test"
    />
  </div>
</tmpage-stub>
`;

exports[`PageProposal shows an error if the proposal couldn't be found 1`] = `
<tmpage-stub
  data-title="Proposal"
  emptysubtitle=""
  emptytitle="No data"
  loaderpath=""
  proposalid="666"
>
  <tmdatanotfound-stub />
</tmpage-stub>
`;
